lib/sinon/test.js

```
@@ -55,6 +55,15 @@
                     sandbox.verifyAndRestore();
                 }
 
+                if (result && typeof result.then === "function" && typeof result.catch === "function") {
```

> If the user supplied a "done" callback and also returned a Promise, how will that case be handled then?

That case should throw an error (like Mocha). The user is giving mixed signals.